### ISSUE TITLE
IAM Roles for Tasks

### ISSUE THREAD

**Issue by ollietreend** [danilop/yas3fs#135]
Hi,

I'm planning to use `yas3fs` inside a docker container, which will be run on top of an [ECS](https://aws.amazon.com/ecs/) cluster. I can see from the README that `yas3fs` supports [IAM Roles for EC2](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html). Whilst I'm not familiar with IAM Roles for EC2, I know that there is an equivalent for docker containers running in ECS – namely [IAM Roles for Tasks](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html).

From reading the docs, it looks like there are some differences between the two implementations which means that the AWS credentials have to be retrieved slightly differently. Therefore, the existing support for IAM roles which is baked in to `yas3fs` is not compatible with the ECS implementation of IAM roles.

As a workaround I have considered implementing the retrieval of the AWS credentials in my init script, prior to starting `yas3fs`. In essence, my plan would be to retrieve the credentials, from which I will be able to pass them in as regular `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables – and `yas3fs` would work as normal.

However the problem I'm coming up against is the fact that the AWS credentials supplied by IAM Roles are temporary, and will expire after just a few hours. Therefore, I would have to implement some mechanism to kill and restart my `yas3fs` process with renewed AWS credentials, prior to the old credentials expiring. This additional management overhead will be clunky and reliant on another process to 'remember' to refresh the credentials – it would be much simpler and less error prone for this to be baked in to `yas3fs`.

Are there any plans to add support for IAM Roles for Tasks?

If not, do you have any suggestions for other workarounds? Unless I can figure something out, I'll have to end up resorting to using long-lived IAM users rather than the magic of IAM roles.

Thanks

**Comment by dacut** [comment]
yas3fs uses the regular Boto 2 library under the hood which has this support baked in.

If you don't specify any credentials (and keep Boto from finding them in the filesystem), Boto will grab the EC2 role credentials from the metadata endpoint automatically (and properly refreshing them when needed). In other words, you shouldn't need to do anything here as long as you've launched your EC2 instance with a properly configured IAM role.

In fact, passing `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` from the metadata endpoint wouldn't be sufficient. You would also need the token code, passed in `AWS_SECURITY_TOKEN`.

The [Boto 2 documentation](http://boto.cloudhackers.com/en/latest/boto_config_tut.html#credentials) is incorrect in this regard. The [Boto 3 documentation](http://boto3.readthedocs.io/en/latest/guide/configuration.html#configuring-credentials) is correct for both versions of the library.

The gory details in Boto 2 can be found in [boto/provider.py](https://github.com/boto/boto/blob/develop/boto/provider.py#L278), starting around line 278. It finally gets to the metadata endpoint [at line 375](https://github.com/boto/boto/blob/develop/boto/provider.py#L278).

**Comment by ollietreend** [comment]
Thanks @dacut – sorry I missed your reply, I've only just seen this.

[IAM Roles for Tasks](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html) (used by the Elastic Container Service) are different to [IAM Roles for EC2](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html).

They essentially do the same thing, but are implemented differently. So as it stands yas3fs does not support IAM Roles for Tasks.

Support [was added in boto3](https://github.com/boto/boto3/issues/713).

So it looks like it'll be necessary to switch to boto3 to gain support for this. Which sounds like a pretty big job – but as somebody who's not experienced with python, I don't know for sure.

**Comment by jyaworski** [comment]
I'm running into something similar on an EC2 instance:

````
$ yas3fs --queue blah --region ap-northeast-1 --s3-use-sigv4 --topic arn:aws:sns:ap-northeast-1:qqqq:blah --mkdir --aws-managed-encryption --prefetch --s3-endpoint https://bukkit.s3-ap-northeast-1.amazonaws.com s3://bukkit testbucket

...

MainThread WARNING Failed to get current user principalId: BotoServerError: 400 Bad Request
<ErrorResponse xmlns="https://iam.amazonaws.com/doc/2010-05-08/">
  <Error>
    <Type>Sender</Type>
    <Code>ValidationError</Code>
    <Message>Must specify userName when calling with non-User credentials</Message>
  </Error>
  <RequestId>0dddddddddd</RequestId>
</ErrorResponse>
````

[This link](https://forums.aws.amazon.com/thread.jspa?threadID=222843) says to make use of `STS.Client.get_caller_identity`, which produced something for me when I ran it:

````
>>> import boto3
>>> boto3.client('sts').get_caller_identity()
{u'Account': 'ffffffff', u'UserId': 'ffffff:i-000000000', 'ResponseMetadata': {'RetryAttempts': 0, 'HTTPStatusCode': 200, 'RequestId': 'ddwf2ef2', 'HTTPHeaders': {'x-amzn-requestid': 'wwddasda', 'date': 'Thu, 11 Jan 2018 15:50:57 GMT', 'content-length': '460', 'content-type': 'text/xml'}}, u'Arn': 'arn:aws:sts::22222:assumed-role/Web-Cluster-EC2/i-0000000'}
````

**Comment by jazzl0ver** [comment]
While my pull request is checking, you may use the version from my repo: https://github.com/jazzl0ver/yas3fs. It's fixed there.

**Comment by omkark95** [comment]
> While my pull request is checking, you may use the version from my repo: https://github.com/jazzl0ver/yas3fs. It's fixed there.

This worked like a charm. Thanks!

**Comment by Aeroxander** [comment]
@jyaworski I have the exact same error, did you find a fix?

**Comment by jazzl0ver** [comment]
this should have been fixed. what version are you using?

**Comment by Aeroxander** [comment]
@jazzl0ver It should be yas3fs 2.3.5, I just started using it a month ago. Locally I don't have this issue but on AWS Batch (what uses ECS) I do.

**Comment by jazzl0ver** [comment]
the latest version is 2.4.6

**Comment by Aeroxander** [comment]
Ah ye I just remember I got an error during the install, it seems like it install an older version then.
Think I have this issue https://github.com/danilop/yas3fs/issues/36

**Comment by jazzl0ver** [comment]
you need to use python2.7

**Comment by Aeroxander** [comment]
Actually I do, I just made it work with installing setuptools:

> pip2 install setuptools --upgrade && \
> 	pip2 install yas3fs 

pip just installs an older version..

**Comment by jazzl0ver** [comment]
Yeah, I've requested access to fix it long time ago (https://github.com/danilop/yas3fs/issues/140) - but nobody cares :(
